### PR TITLE
Fix shiny Tileset when using KHR Materials Unlit

### LINKED ISSUE #1055: Lighting for `KHR_materials_unlit` models is weird when the sun is near the horizon
Our technique for rendering unlit models (use constant normals aligned with the globe up direction) creates some weird artifacts when using Unreal's SunSky and the sun is near the horizon. For example, see this forum post:
https://community.cesium.com/t/about-cesiumsunsky-light-error-through-of-3dtiles-question/22692

image

It's not too obvious to me what to do about this. For starters, we should probably at least provide the option to ignore `KHR_materials_unlit`.

### PR DESCRIPTION
**Pull request by joseph-kaile**
Fixes #1055

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @joseph-kaile!

* :x: Missing CLA.
    * Please send in a [Contributor License Agreement](https://github.com/CesiumGS/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla) (CLA) and comment back here to let us know to check this!
    * Maintainers, [here's how to check](https://github.com/CesiumGS/cesium/blob/master/Documentation/Contributors/CLAs/README.md) for a submitted CLA.

Reviewers, don't forget to make sure that:

- [ ] [Cesium for Unreal Samples](https://github.com/CesiumGS/cesium-unreal-samples) works.

File: CHANGES.md [hunk 1]
```diff
@@ -6,6 +6,10 @@
 
 - The `FlyToAltitudeProfileCurve`, `FlyToProgressCurve`, `FlyToMaximumAltitudeCurve`, `FlyToDuration`, and `FlyToGranularityDegrees` properties of `GlobeAwareDefaultPawn`  / `DynamicPawn` may now be read and written from Blueprints.
 
+##### Fixes :wrench:
+
+- Fixed wierd reflection on tilesets with KHR Materials Unlit when the sun is close to the horizon. Added option to ignore KHR Materials Unlit.
```
Comment (kring) [inline CHANGES.md:11]:
```suggestion
- Fixed unexpected reflection on tilesets with `KHR_materials_unlit` extension when the sun is close to the horizon. 
- Added an option to ignore the `KHR_materials_unlit` extension entirely.
```

File: Source/CesiumRuntime/Private/CesiumGltfComponent.cpp [hunk 2]
```diff
@@ -779,10 +779,26 @@ static void loadPrimitive(
...
       materialID >= 0 && materialID < model.materials.size()
           ? model.materials[materialID]
           : defaultMaterial;
+
+  bool isUnlit = material.hasExtension<ExtensionKhrMaterialsUnlit>() &&
+                 !options.pMeshOptions->pNodeOptions->pModelOptions
+                      ->ignoreKhrMaterialsUnlit;
+  if (isUnlit) {
+    Material& unlitMaterial = const_cast<Material&>(material);
+    unlitMaterial.emissiveFactor = {1.0, 1.0, 1.0};
```
Comment (kring) [inline Source/CesiumRuntime/Private/CesiumGltfComponent.cpp:793]: Why should the unlit material have an emissiveFactor of 1.0? I don't quite understand the logic here.

File: Source/CesiumRuntime/Private/CesiumGltfComponent.cpp [hunk 3]
```diff
@@ -779,10 +779,26 @@ static void loadPrimitive(
...
   const Material& material =
       materialID >= 0 && materialID < model.materials.size()
           ? model.materials[materialID]
           : defaultMaterial;
+
+  bool isUnlit = material.hasExtension<ExtensionKhrMaterialsUnlit>() &&
+                 !options.pMeshOptions->pNodeOptions->pModelOptions
+                      ->ignoreKhrMaterialsUnlit;
+  if (isUnlit) {
+    Material& unlitMaterial = const_cast<Material&>(material);
```
Comment (kring) [inline Source/CesiumRuntime/Private/CesiumGltfComponent.cpp:792]: Can we avoid modifying the material? It's pretty dodgy for a function to take a const parameter, then cast away that const and modify it anyway.

File: Source/CesiumRuntime/Private/CesiumGltfComponent.cpp [hunk 4]
```diff
@@ -1913,6 +1929,13 @@ static void loadPrimitiveGameThreadPart(
   pMesh->SetCustomDepthStencilValue(
       pGltf->CustomDepthParameters.CustomDepthStencilValue);
 
+  static constexpr std::array<double, 3> emissiveFactor = {1.0, 1.0, 1.0};
+  if (loadResult.pMaterial && std::equal(
+                                  loadResult.pMaterial->emissiveFactor.begin(),
+                                  loadResult.pMaterial->emissiveFactor.end(),
+                                  emissiveFactor.begin())) {
+    pMesh->bCastDynamicShadow = false;
```
Comment (kring) [inline Source/CesiumRuntime/Private/CesiumGltfComponent.cpp:1937]: Are you sure about this? It's possible for a (normal, not unlit) material to both have an emissive component, _and_ the user expects it to cast shadows like normal, right?

**Review by kring — COMMENTED** [review]
Just a few questions @joseph-kaile. This might be fine, but I don't totally understand the rationale for some of the choices here. It needs some more explanation, at least.

**Comment by joseph-kaile** [comment]
@kring , thanks for the suggestions. The PR is now improved.

**Comment by kring** [comment]
I gave the new property a better display name, and added some doc for it. Otherwise, looks great! Big visual improvement. Merging.

**Comment by GunOmniscience** [comment]
Good job！！！😁👍

**Pull request merged**